### PR TITLE
Struct mech/multi stage expl

### PR DESCRIPTION
**Pull request by KlausBSautter**
I added a new explicit time integration strategy based on:

> An accurate two‐stage explicit time integration scheme for structural dynamics and various dynamic problems

> Wooram Kim

> https://onlinelibrary.wiley.com/doi/abs/10.1002/nme.6098

@loumalouomega I had to rearrange the strategy a little bit. Can we discuss this? Is this ok for you?

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
We can discuss of course. I don't see anything wrong to begin with. I will check the paper and revisit the PR.

Just one thing, please add tests (you can extend current explicit tests with the proper specialisation using this scheme)

Conclusion : Add tests please

**Comment by KlausBSautter** [comment]
> I will check the paper and revisit the PR.

perfect, thanks

**Comment by loumalouomega** [comment]
I checked the paper and added the test. Is much more stable, congrats, I will approve after adding some minor comments.

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:341]: why is this removed?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_explicit_dynamic_solver.py
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_explicit_dynamic_solver.py]: 1. Is this needed to be a historical variable?
2. If yes then please add it only if the "multi_stage" scheme is used

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp]:
```suggestion
//  Main authors:    Klaus B Sautter
```
this is your work isn't it?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp:232]: is this different from the version in the core?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp:275]: this will segfault if for some reason there are no nodes

**Comment by KlausBSautter** [comment]
as soon as #5121 is merged, I'd like to delete `InitializeNonLinIteration`

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp]: ?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp]: of this is the only call then you can remove this

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp:488]: usually I would recommend to pass a ref to the node unless you need the iterator

**Review by philbucher — CHANGES_REQUESTED** [review]
some minor changes required, otherwise nice work!

**Review by loumalouomega — DISMISSED** [review]
Now OK

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:341]: this is now in the scheme. I need to call it multiple times in the new scheme.

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp:232]: no, there is a pending PR  : #5121

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp:488]: I think we need the iterator here, just like we do in the central differencing

**Comment by KlausBSautter** [comment]
@philbucher is it ok like that?

**Comment by philbucher** [comment]
@KlausBSautter lets go through it together on Monday, I think this is more effective than writing everything here (i only have some minor comments)

**Review by philbucher — APPROVED** [review]
ok after discussing in person

**Pull request merged**
